examples/print.js

```diff
@@ -0,0 +1,248 @@
+   */
+  this['map'] = new ol.Map({
+    layers: [
+      new ol.layer.Image({
+        source: new ol.source.ImageWMS({
+          url: app.WMS_URL_,
+          params: {
+            'LAYERS': 'osm'
+          },
+          serverType: /** @type {ol.source.wms.ServerType} */ ('mapserver')
+        })
+      })
+    ],
+    view: new ol.View({
+      projection: projection,
+      resolutions: [50, 20, 10, 5, 2.5, 2, 1],
+      center: [537635, 152640],
+      zoom: 3
+    })
+  });
+
+  /**
+   * Text to display a "loading" message while waiting for the report.
+   * @type {string}
+   */
+  this['printState'] = '';
+
+  /**
+   * @type {angular.$timeout}
+   * @private
+   */
+  this.$timeout_ = $timeout;
+
+  /**
+   * @type {ngeo.Print}
+   * @private
+   */
+  this.print_ = ngeoCreatePrint(app.PRINT_URL_);
+
+};
+
+
+/**
+ * @private
+ * @param {ol.Map} map Map.
+ * @return {number} The optimal scale.
+ */
+app.MainController.getOptimalScale_ = function(map) {
+
+  var mapSize = map.getSize();
+  var resolution = map.getView().getResolution();
+  var mapWidth = mapSize[0] * resolution;
+  var mapHeight = mapSize[1] * resolution;
+
+  var scaleWidth = mapWidth * app.INCHES_PER_METER_ *
+      app.DOTS_PER_INCH_ / app.PRINT_PAPER_SIZE_[0];
+  var scaleHeight = mapHeight * app.INCHES_PER_METER_ *
+      app.DOTS_PER_INCH_ / app.PRINT_PAPER_SIZE_[1];
+
+  var pageScale = Math.min(scaleWidth, scaleHeight);
+  var optimal = Infinity;
+
+  var i, ii;
+  for (i = 0, ii = app.PRINT_SCALES_.length; i < ii; i++) {
+    if (pageScale > app.PRINT_SCALES_[i]) {
+      optimal = app.PRINT_SCALES_[i];
+    }
+  }
+
+  return optimal;
+};
+
+
+/**
+ * @export
+ */
+app.MainController.prototype.print = function() {
+  var map = this['map'];
+
+  var scale = app.MainController.getOptimalScale_(map);
+  var dpi = app.PRINT_DPI_;
+  var layout = app.PRINT_LAYOUT_;
+
+  this['printState'] = 'Printing…';
+
+  var spec = this.print_.createSpec(map, scale, dpi, layout, {
+    'datasource': []
+  });
+
+  this.print_.createReport(spec).then(
+      angular.bind(this, this.handleCreateReportSuccess_),
+      angular.bind(this, this.handleCreateReportError_));
+};
+
+
+/**
+ * @param {Object} data Data.
+ * @private
+ */
+app.MainController.prototype.handleCreateReportSuccess_ = function(data) {
+  this.getStatus_(data['ref']);
+};
+
+
+/**
+ * @param {string} ref Ref.
+ * @private
+ */
+app.MainController.prototype.getStatus_ = function(ref) {
+  this.print_.getStatus(ref).then(
+      angular.bind(this, this.handleGetStatusSuccess_, ref),
+      angular.bind(this, this.handleGetStatusError_));
+};
+
+
+/**
+ * @param {Object} data Data.
+ * @private
+ */
+app.MainController.prototype.handleCreateReportError_ = function(data) {
+  this['printState'] = 'Print error';
+};
+
+
+/**
+ * @param {string} ref Ref.
+ * @param {Object} data Data.
+ * @private
+ */
+app.MainController.prototype.handleGetStatusSuccess_ = function(ref, data) {
+  var done = data['done'];
```

Comment:
We need new types in the mapfish-print-v3.js externs file for the print responses.